File: src/time_utils.py
import os
import re
import googlemaps
from datetime import datetime

from input_utils import prompt_for_event_time
from text_utils import print_red, print_stats


from datetime import datetime


def get_event_time(event_time: str | None) -> datetime | None:
    """Get the event time as a datetime object."""

    if event_time is None:
        print_red("No event time provided.")
        event_time = prompt_for_event_time()

    try:
        return datetime.strptime(event_time, "%I:%M %p")
    except ValueError:
        print_red(f"Invalid event time: {event_time}. Please re-enter.")
        event_time = prompt_for_event_time()
        try:
            return datetime.strptime(event_time, "%I:%M %p")
        except ValueError:
            return None


def get_directions(origin: str, destination: str):
    """Fetch directions using Google Maps API."""

    gmaps = googlemaps.Client(key=os.getenv("GOOGLE_MAPS_API_KEY"))
    try:
        return gmaps.directions(origin, destination, departure_time=datetime.now())  # type: ignore
    except Exception as e:
        print_red(f"Error fetching directions: {e}")
        return None


def parse_duration_and_distance(directions_result):
    """Extract duration and distance from the directions result."""

    if not directions_result:
        print_red("No directions found.")
        return None, None

    leg = directions_result[0]["legs"][0]
    duration_text = leg["duration"]["text"]
    distance_text = leg["distance"]["text"]

    return duration_text, distance_text


def extract_time_from_text(duration_text: str) -> int:
    """Extract total time in minutes from duration text."""

    hours = minutes = 0
    if hours_match := re.search(r"(\d+)\s*hour", duration_text):
        hours = int(hours_match.group(1))
    if minutes_match := re.search(r"(\d+)\s*min", duration_text):
        minutes = int(minutes_match.group(1))
    return hours * 60 + minutes


def extract_distance_and_unit(distance_text: str):
    """Extract distance and unit from distance text."""

    if distance_match := re.search(r"(\d+(?:\.\d+)?)\s*(km|mi)", distance_text):
        distance = float(distance_match.group(1))
        unit = distance_match.group(2)
        return distance, unit
    return None, None


def get_base_travel_time(origin: str, destination: str) -> int | None:
    """Get the base travel time based on the origin and destination."""

    directions_result = get_directions(origin, destination)
    duration_text, distance_text = parse_duration_and_distance(directions_result)
    if duration_text is None or distance_text is None:
        return None

    total_minutes = extract_time_from_text(duration_text)
    distance, unit = extract_distance_and_unit(distance_text)
    if distance and unit:
        roundtrip_distance = distance * 2
        print_stats(f"Total roundtrip mileage: {roundtrip_distance} {unit}\n")

    return total_minutes

File: src/travel/time_utils.go
package travel

import (
	"context"
	"errors"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"time"

	"github.com/fatih/color"
	"github.com/jlsnow301/cutsheet-timer/input"
	"googlemaps.github.io/maps"
)

// getEventTime gets the event time as a time.Time object.
func GetEventTime(eventTime string) (*time.Time, error) {
	if eventTime == "" {
		color.Red("No event time provided.")
		eventTime = input.PromptForEventTime()
	}

	parsedTime, err := time.Parse("03:04 PM", eventTime)
	if err != nil {
		color.Red(fmt.Sprintf("Invalid event time: %s. Please re-enter.", eventTime))
		eventTime = input.PromptForEventTime()
		parsedTime, err = time.Parse("03:04 PM", eventTime)
		if err != nil {
			return nil, err
		}
	}
	return &parsedTime, nil
}

// getDirections fetches directions using Google Maps API.
func getDirections(origin, destination string) *maps.Route {
	client, err := maps.NewClient(maps.WithAPIKey(os.Getenv("GOOGLE_MAPS_API_KEY")))
	if err != nil {
		color.Red(fmt.Sprintf("Error creating Google Maps client: %v", err))
		return nil
	}

	request := &maps.DirectionsRequest{
		Origin:        origin,
		Destination:   destination,
		DepartureTime: "now",
	}

	routes, _, err := client.Directions(context.Background(), request)
	if err != nil {
		color.Red(fmt.Sprintf("Error fetching directions: %v", err))
		return nil
	}

	if len(routes) > 0 {
		return &routes[0]
	}
	return nil
}

// parseDurationAndDistance extracts duration and distance from the directions result.
func parseDurationAndDistance(directionsResult *maps.Route) (string, string) {
	if directionsResult == nil {
		color.Red("No directions found.")
		return "", ""
	}

	leg := directionsResult.Legs[0]
	durationText := leg.Duration.String()
	distanceText := leg.Distance.HumanReadable

	return durationText, distanceText
}

// extractTimeFromText extracts total time in minutes from duration text.
func extractTimeFromText(durationText string) int {
	hours, minutes := 0, 0
	hoursMatch := regexp.MustCompile(`(\d+)\s*hour`).FindStringSubmatch(durationText)
	if len(hoursMatch) > 1 {
		hours, _ = strconv.Atoi(hoursMatch[1])
	}
	minutesMatch := regexp.MustCompile(`(\d+)\s*min`).FindStringSubmatch(durationText)
	if len(minutesMatch) > 1 {
		minutes, _ = strconv.Atoi(minutesMatch[1])
	}
	return hours*60 + minutes
}

// extractDistanceAndUnit extracts distance and unit from distance text.
func extractDistanceAndUnit(distanceText string) (float64, string) {
	distanceMatch := regexp.MustCompile(`(\d+(?:\.\d+)?)\s*(km|mi)`).FindStringSubmatch(distanceText)
	if len(distanceMatch) > 2 {
		distance, _ := strconv.ParseFloat(distanceMatch[1], 64)
		unit := distanceMatch[2]
		return distance, unit
	}
	return 0, ""
}

// GetBaseTravelTime gets the base travel time based on the origin and destination.
func GetBaseTravelTime(origin, destination string) (*int, error) {
	directionsResult := getDirections(origin, destination)
	durationText, distanceText := parseDurationAndDistance(directionsResult)
	if durationText == "" || distanceText == "" {
		return nil, errors.New("no directions found")
	}

	totalMinutes := extractTimeFromText(durationText)
	distance, unit := extractDistanceAndUnit(distanceText)
	if distance > 0 && unit != "" {
		roundtripDistance := distance * 2
		color.Green(fmt.Sprintf("Total roundtrip mileage: %.2f %s\n", roundtripDistance, unit))
	}

	return &totalMinutes, nil
}
